### PR TITLE
Add serviceable property to lock file

### PR DESCRIPTION
**Pull request by ChengTian**
parent https://github.com/aspnet/dnx/issues/1425 https://github.com/aspnet/dnx/issues/1097

@davidfowl @victorhurdugaci @lodejard @glennc 

Did verification with the ServiceableLib provided by @glennc

### CODE REVIEW

File: src/Microsoft.Framework.Runtime/DependencyManagement/LockFileLibrary.cs [hunk 1]
```diff
@@ -14,8 +14,11 @@ public class LockFileLibrary
 
         public SemanticVersion Version { get; set; }
 
+        public bool Serviceable { get; set; }
```
Comment (victorhurdugaci) [inline src/Microsoft.Framework.Runtime/DependencyManagement/LockFileLibrary.cs]: IsServiceable

File: src/Microsoft.Framework.PackageManager/Utils/LockFileUtils.cs [hunk 2]
```diff
@@ -129,5 +146,97 @@ private static List<string> GetPackageAssemblies(IPackage package, FrameworkName
             return results;
         }
 
+        private static bool IsAssemblyServiceable(string assemblyPath)
+        {
+            using (var stream = File.OpenRead(assemblyPath))
```
Comment (victorhurdugaci) [inline src/Microsoft.Framework.PackageManager/Utils/LockFileUtils.cs]: Check that the file exists

File: src/Microsoft.Framework.Runtime/DependencyManagement/LockFileLibrary.cs [hunk 3]
```diff
@@ -14,8 +14,11 @@ public class LockFileLibrary
...
+
         public string Sha { get; set; }
 
+
```
Comment (victorhurdugaci) [inline src/Microsoft.Framework.Runtime/DependencyManagement/LockFileLibrary.cs]: Empty line...

File: src/Microsoft.Framework.PackageManager/Utils/LockFileUtils.cs [hunk 4]
```diff
@@ -129,5 +146,97 @@ private static List<string> GetPackageAssemblies(IPackage package, FrameworkName
...
+                    var container = mdReader.GetMemberReference((MemberReferenceHandle)ctorHandle).Parent;
+                    var name = mdReader.GetTypeReference((TypeReferenceHandle)container).Name;
+                    if (!string.Equals(mdReader.GetString(name), "AssemblyMetadataAttribute"))
+                    {
+                        continue;
+                    }
+
+                    var arguments = GetFixedStringArguments(mdReader, attr);
+                    if (arguments.Count == 2 &&
+                        string.Equals(arguments[0], "Serviceable", StringComparison.OrdinalIgnoreCase) &&
```
Comment (victorhurdugaci) [inline src/Microsoft.Framework.PackageManager/Utils/LockFileUtils.cs]: `ServiceableAttribute` and `serviceableAttribute` are two different types. It shouldn't be case insensitive

File: src/Microsoft.Framework.PackageManager/Utils/LockFileUtils.cs [hunk 5]
```diff
@@ -129,5 +146,97 @@ private static List<string> GetPackageAssemblies(IPackage package, FrameworkName
...
+                        string.Equals(arguments[1], "True", StringComparison.OrdinalIgnoreCase))
```
Comment (victorhurdugaci) [inline src/Microsoft.Framework.PackageManager/Utils/LockFileUtils.cs]: Same here (as above)?

**Comment by victorhurdugaci** [comment]
You will need to edit the breadcrumbs checker to look for the `IsServiceable` property

**Comment by victorhurdugaci** [comment]
Btw, can you add 2 tests (one with an assembly that has the attribute and one with an assembly without the attribute)?

File: src/Microsoft.Framework.PackageManager/Utils/LockFileUtils.cs (hunk 4 above)
Comment (ChengTian) [inline src/Microsoft.Framework.PackageManager/Utils/LockFileUtils.cs]: Actually that's the key of metadata, not a type name:

serviceable

@glennc , is the metadata key and value case-sensitive? Do we accept
```
[assembly: AssemblyMetadata("serviceable", "true")]
```
in addition to
```
[assembly: AssemblyMetadata("Serviceable", "True")]
```

**Comment by ChengTian** [comment]
Breadcrumbs checker was added before I did verification. Seems my force push failed so that change didn't show here. Please check the updated commit.

File: src/Microsoft.Framework.Runtime/Servicing/Breadcrumbs.cs [hunk 6]
```diff
@@ -33,8 +33,7 @@ public Breadcrumbs(string breadcrumbsFolder)
 
         public static bool IsPackageServiceable(PackageInfo package)
         {
-            // TODO: Figure out what makes a package serviceable
-            return false;
+            return package.LockFileLibrary.Serviceable;
```
Comment (victorhurdugaci) [inline src/Microsoft.Framework.Runtime/Servicing/Breadcrumbs.cs]: Either your or me will have to rebase (depending on which change goes first) :)

File: src/Microsoft.Framework.PackageManager/Utils/LockFileUtils.cs (hunk 4 above)
Comment (glennc) [inline src/Microsoft.Framework.PackageManager/Utils/LockFileUtils.cs]: Leave it as case insensitive for now.

**Comment by ChengTian** [comment]
Added tests. Any additional comments?

**Comment by victorhurdugaci** [comment]
Did you test the reader on coreclr? If that works, then :shipit: from me

File: src/Microsoft.Framework.PackageManager/Utils/LockFileUtils.cs [hunk 7]
```diff
@@ -129,5 +146,102 @@ private static List<string> GetPackageAssemblies(IPackage package, FrameworkName
...
+
+        /// <summary>
+        /// Gets the fixed (required) string arguments of a custom attribute.
+        /// Only attributes that have only fixed string arguments.
+        /// </summary>
+        private static List<string> GetFixedStringArguments(MetadataReader reader, CustomAttribute attribute)
+        {
+            // TODO: Nick Guerrera (Nick.Guerrera@microsoft.com) hacked this method for temporary use.
+            // There is a blob decoder feature in progress but it won't ship in time for our milestone.
+            // Replace this method with the blob decoder feature when later it is availale.
```
Comment (victorhurdugaci) [inline src/Microsoft.Framework.PackageManager/Utils/LockFileUtils.cs]: Add a bug for this so we don't forget about the hacky code

**Comment by ChengTian** [comment]
Did verification with CoreCLR and Mono. Thanks for the sign-off.

**Pull request merged**